### PR TITLE
Fixes for recent darwin.js errors & enhancements to per generation output

### LINKED ISSUE #872: Cannot read property 'vsBuyHold'
After running first generation in Darwin the following error appears:
 
**TypeError: Cannot read property 'vsBuyHold' of undefined**
       at Object.fitness [as fitnessFunction] 
   (/home/geronimo/zenbot/scripts/genetic_backtester/phenotype.js:66:40)
       at /home/geronimo/zenbot/node_modules/geneticalgorithm/index.js:135:38
       at Array.map (<anonymous>)

Have been checking out and it appears that darwin isn't storing any data in it's CSV and JSON files.
i.e.  backtesting_1513035564.csv and generation_data_1513035564_gen_2.json

After modifying line 551 in darwin.js from:
`_fs.writeFile(fileName, csv, err => {_`
to:
`_fs.writeFileSync(fileName, csv, err => {_`

it has started storing data in CSV and JSON files but the error still remains. Since it's a parallel execution I suspect that it's trying to get the variable before any data is stored in files.

Any suggestions ?

### PR DESCRIPTION
**Pull request by tiagosiebler**
- best.sim was oddly undefined every generation I tried. This started in a recent PR (unsure which), so instead I’m pulling this from the first row in the ordered results. Seems to be working. This fixes #872 for now, until someone figures out where .sim was supposed to come from in the `best` object (line 623). I’ve kept the original way, in case it isn’t broken for everyone (and somehow causes issues?)
- Refactored the json & csv generation logic, so the next generation won’t start until saving finishes. Previously, darwin would misleadingly start the next generation before saving completes, often resulting in backtesting being killed too early before the files actually saved, since the assumption is if the next gen started, surely the last is finished and saved?
- Added messages when saving to file completes, including file-names (for easier resuming of generations later).
- Added additional output each generation. This includes top command from that generation, ready to copy & paste, as well as additional trade stats for those results.

Example output between generations:

> Generation complete, saving results...
> > Finished writing generation csv to simulations/backtesting_1514395796.csv
> > Finished writing generation json to simulations/generation_data_1514395796_gen_2.json
> 
> 
> Generation's Best Results
> 	(crossover_vwap) Result Fitness -32.35, VS Buy and Hold: -32.35, End Balance: 1000.5002935, Wins/Losses 0/0.
> ./zenbot.sh sim kraken.XXRP-ZEUR --avg_slippage_pct=0.045 --buy_stop_pct=0 --emalen1=251 --markdown_buy_pct=3.7463910526589608 --markup_pct=undefined --markup_sell_pct=-0.15815181265159195 --max_sell_loss_pct=25 --max_slippage_pct=5 --min_periods=118 --order_type=maker --period=388m --periodLength=388m --profit_stop_enable_pct=19 --profit_stop_pct=11 --rsi_periods=14 --sell_stop_pct=40 --smalen1=108 --smalen2=60 --strategy=crossover_vwap --vwap_length=255 --vwap_max=0 --days=7 --asset_capital=0 --currency_capital=1000
> 
> 
> 
> === Simulating generation 2 ===
> 
> Backfilling (might take some time) ...

### CODE REVIEW

**Review by davobo — APPROVED** [review]
This fix solves the issue reported in bug 872.

File: scripts/genetic_backtester/darwin.js [hunk 1]
```diff
@@ -560,44 +596,51 @@ let simulateGeneration = () => {
...
+    saveGenerationData(csvFileName, jsonFileName, dataCSV, dataJSON, (id)=>{
+      filesSaved++;
+      if(filesSaved == 2){        
+        console.log(`\n\nGeneration's Best Results`);
+        selectedStrategies.forEach((v)=> {
+          let best = pools[v]['pool'].best();      
+          if(best.sim)
+            console.log(`\t(${v}) Sim Fitness ${best.sim.fitness}, VS Buy and Hold: ${best.sim.vsBuyHold} End Balance: ${best.sim.endBalance}, Wins/Losses ${best.sim.wins}/${best.sim.losses}.`);
+          else 
+            console.log(`\t(${v}) Result Fitness ${results[0].vsBuyHold}, VS Buy and Hold: ${results[0].vsBuyHold}, End Balance: ${results[0].endBalance}, Wins/Losses ${results[0].wins}/${results[0].losses}.`);
```
Comment (evseevnn-zz) [inline scripts/genetic_backtester/darwin.js]:
```
Result Fitness ${results[0].vsBuyHold}, VS Buy and Hold: ${results[0].vsBuyHold}
```
should be
```
Result Fitness ${results[0].fitness}, VS Buy and Hold: ${results[0].vsBuyHold}
```

File: scripts/genetic_backtester/darwin.js [hunk 2]
```diff
@@ -560,44 +596,51 @@ let simulateGeneration = () => {
...
+    let jsonFileName = `simulations/generation_data_${fileDate}_gen_${generationCount}.json`;
+    let dataJSON = JSON.stringify(poolData, null, 2);
+    var filesSaved = 0;
+    saveGenerationData(csvFileName, jsonFileName, dataCSV, dataJSON, (id)=>{
+      filesSaved++;
+      if(filesSaved == 2){        
+        console.log(`\n\nGeneration's Best Results`);
+        selectedStrategies.forEach((v)=> {
+          let best = pools[v]['pool'].best();      
+          if(best.sim)
```
Comment (evseevnn-zz) [inline scripts/genetic_backtester/darwin.js]: Please use syntax with `{`

**Review by evseevnn-zz — COMMENTED** [review]
@tiagosiebler Thank for your work.
Please fix small things.

File: scripts/genetic_backtester/darwin.js (hunk 1 above)
Comment (tiagosiebler) [reply scripts/genetic_backtester/darwin.js]: Good catch! Can't believe I missed that...

**Comment by tiagosiebler** [comment]
Added two edits based on @evseevnn's comments. If anything else is needed, let me know.

**Review by DeviaVir — APPROVED** [review]

**Pull request merged**